### ISSUE TITLE
Trezor watching-only mode logic was broken after keepkey code combine?

### ISSUE THREAD

**Issue by mmortal03** [spesmilo/electrum#1845]
When using Electrum with a Trezor, we used to be able to load Electrum *without* the Trezor plugged in, and Electrum would simply tell the user, "Trezor device not detected. Continuing in watching-only mode."

Now, it doesn't do this, it prompts the user with, "Could not connect to your TREZOR.  Verify the cable is connected and that no other application is using it. Try to connect again?"

If the user chooses No, it asks again. If the user chooses No again, it doesn't prompt a third time (however, twice is already too much), but then will later pop up again after a period of time.

This was removed: 

`self.wallet.force_watching_only = True`

(https://github.com/spesmilo/electrum/commit/3a1bb5b61a5126ad47ff01a609b0e24db800c1f5#diff-c098321b72b58006649f7de98383acf1L103)

and this:
```
    def is_watching_only(self):		
         return self.force_watching_only
```

(https://github.com/spesmilo/electrum/commit/3a1bb5b61a5126ad47ff01a609b0e24db800c1f5#diff-bdf5807738d9c7d16b1e98009eea7fa0L76)

became:
```
def is_watching_only(self):
         return self.checked_device and not self.proper_device
```
(https://github.com/spesmilo/electrum/commit/3a1bb5b61a5126ad47ff01a609b0e24db800c1f5#diff-9a32635d4a93b6e41c91fec9e4001d4aR1740)

Might this new logic be causing "is_watching_only(self)" to no longer return True whenever a Trezor isn't plugged in?

The new code became:

```
        msg = self.wallet.sanity_check()
        if msg:
            window.show_error(msg)
```

which runs the following function, but sanity_check()'s try exception doesn't seem to be reached in this scenario (which may just be a symptom of watching-only not being handled correctly in the first place):

```
    def sanity_check(self):
         try:
             self.get_client().ping('t')
         except BaseException as e:
             return _("%s device not detected.  Continuing in watching-only "
                      "mode.") % self.device + "\n\n" + str(e)

```

(https://github.com/spesmilo/electrum/commit/3a1bb5b61a5126ad47ff01a609b0e24db800c1f5#diff-9a32635d4a93b6e41c91fec9e4001d4aR1868)

I don't know the code very well, so it'd be great if someone could take a look at this issue.

**Comment by kyuupichan** [comment]
It doesn't go into watching only mode, precisely so you can use it later. 

I do this all the time, it works perfectly well, and it only asks once.  So the first thing to understand is why you're being asked twice.

**Comment by mmortal03** [comment]
Ok, here's a bit more information. 

I did not mention that it prompts me, at the *very* beginning, after it says "Syncronizing" in the bottom left hand corner, with: 

1.) "All the accounts in your wallet have received transactions. Electrum must check whether more accounts exist; one will only be shown if it has been used or you give it a name." After clicking OK to this  (which is annoying to do each time I start Electrum),...

2.) then it prompts me with:  "Could not connect to your TREZOR. Verify the cable is connected and that no other application is using it. Try to connect again?" Then I click NO, and...

3.) then it prompts me, again, with:  "Could not connect to your TREZOR. Verify the cable is connected and that no other application is using it. Try to connect again?" Then I click NO, and...

4.) then it's fine for a while, then it prompts me all over again with: "All the accounts in your wallet have received transactions. Electrum must check whether more accounts exist; one will only be shown if it has been used or you give it a name." Then I click OK, and...

5.) then it prompts me with:  "Could not connect to your TREZOR. Verify the cable is connected and that no other application is using it. Try to connect again?" Then I click NO, and...

6.) Go to number 4. 

Based on my own logic, I suspect that number 2 may always follow number 1 (which may never be getting resolved), and number 4 and 5 just keep coming back up in succession after some sort of sync event or retry.

Number 3 might be the regular prompt. 

Btw, with the Trezor plugged in, it also gives me prompt number 1, but not number 2 or 3. However, it seems to intermittently freeze after that, and if I click the Trezor icon in the bottom right, it gives me number 2, so I don't know if it's ever really connecting. It's been a while since I've used it.

Further context, the Trezor works fine with wallet.mytrezor.com, so I don't think it's anything wrong with the device. I've tried Electrum with and without trezord running in the background, so I don't think that matters, either.

**Comment by kyuupichan** [comment]
I see; it sounds like the account issue is the real problem here.  You're adding the account and still getting the account prompts?  How many accounts have you got shown in the GUI?

**Comment by mmortal03** [comment]
Just one account, never had any more on the Trezor. Nothing that was set up in a weird way. It worked fine back at the end of 2015 when I last used it. Now I've upgraded Electrum to the latest version, and I'm seeing these issues.

**Comment by kyuupichan** [comment]
OK.  Can you keep your current wallet as-is, and restore the same one to a new file from its seed?  And let me know what the difference between the two files is.  I'm pretty sure the newly created one will be fine; something is confusing electrum about the old wallet file.  Once I know the difference it should be easy to track down / fix.

**Comment by kyuupichan** [comment]
I suspect it is something to do with an osbolete next_account2 entry, and/or the master_public_keys entry.

**Comment by mmortal03** [comment]
I backed up the Electrum data directory and started fresh. Then I selected "Create a new wallet" -> "Hardware wallet" -> "Use a device you have already set up" -> "TREZOR wallet". The install wizard ran for a second, then became unresponsive. I had to force quit it. I tried this multiple times, with both the Electrum portable version, and the regular installed version on the machine. Same result every time (the install wizard becomes unresponsive). 

I could try generating a new address and/or do some small sends with wallet.mytrezor.com, but if that fixed it, we might be left with nothing to reproduce the buggy behavior. Could sending you the wallet file be enough to debug it (it doesn't contain any bitcoins currently), or are you able to provide some steps to help me debug it on my end?

**Comment by ecdsa** [comment]
I just tested this with the latest code, I think the current behaviour is acceptable, so closing this.
reopen if you still have issues.

**Comment by mmortal03** [comment]
@ecdsa , are you referring to the master branch (to become 2.7.0)? I'm still on 2.6.4 with this issue. I could test the latest code, but I haven't tried compiling Electrum myself, so moving to the latest code might complicate things for me. Where this issue left off is that I've not transacted with the wallet file so that someone could try reproducing the issue on their end, or could walk me through debugging the issue on my end.

**Comment by ecdsa** [comment]
I'm always referring to the latest code.

**Comment by mmortal03** [comment]
I've given compiling the latest source code a shot, and, after lots of trial and error, I've finally built the Windows executable from the latest code. However, I can't get the hardware wallet functionality working with the executable, so this makes it impossible for me to test this bug on the latest code. I did run build-wine/prepare-hw.sh before buiding the executables, so that's not the issue. Any help would be appreciated.
